python_modules/dagster/dagster/_core/definitions/multi_dimensional_partitions.py

```diff
@@ -50,3 +60,123 @@ def __new__(cls, partition_dimension_mapping: Mapping[str, str]):
+    ):
+        return super().__new__(
+            cls,
+            name=check.str_param(name, "name"),
+            partitions_def=check.inst_param(partitions_def, "partitions_def", PartitionsDefinition),
+        )
+
+
+@experimental
+class MultiPartitionsDefinition(PartitionsDefinition):
+    """The set of partitions is the cross product of partitions in the inner partitions
+    definitions"""
+
+    def __init__(self, partitions_defs: Mapping[str, PartitionsDefinition]):
+        from dagster import DagsterInvalidInvocationError
+
+        if not len(partitions_defs.keys()) == 2:
+            raise DagsterInvalidInvocationError(
+                "Dagster currently only supports multi-partitions definitions with 2 partitions definitions. "
+                f"Your multi-partitions definition has {len(partitions_defs.keys())} partitions definitions."
+            )
+        check.mapping_param(
+            partitions_defs, "partitions_defs", key_type=str, value_type=PartitionsDefinition
+        )
+
+        self._partitions_defs: List[PartitionDimensionDefinition] = sorted(
+            [
+                PartitionDimensionDefinition(name, partitions_def)
+                for name, partitions_def in partitions_defs.items()
+            ],
+            key=lambda x: x.name,
+        )
+
+    @property
+    def partitions_defs(self) -> Sequence[PartitionDimensionDefinition]:
+        return self._partitions_defs
+
+    def get_partitions(self, current_time: Optional[datetime] = None) -> Sequence[Partition]:
+        partition_sequences = [
+            partition_dim.partitions_def.get_partitions(current_time=current_time)
+            for partition_dim in self._partitions_defs
+        ]
+
+        def get_multi_dimensional_partition(partitions_tuple: Tuple[Partition]):
+            check.invariant(len(partitions_tuple) == len(self._partitions_defs))
+
+            partitions_by_dimension: Dict[str, Partition] = {
+                self._partitions_defs[i].name: partitions_tuple[i]
+                for i in range(len(partitions_tuple))
+            }
+
+            return Partition(
+                value=partitions_by_dimension,
+                name=self.get_partition_key(
+                    {
+                        dimension_key: partition.name
+                        for dimension_key, partition in partitions_by_dimension.items()
+                    }
+                ),
+            )
+
+        return [
+            get_multi_dimensional_partition(partitions_tuple)
+            for partitions_tuple in itertools.product(*partition_sequences)
+        ]
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, MultiPartitionsDefinition)
+            and self.partitions_defs == other.partitions_defs
+        )
+
+    def __hash__(self):
+        return hash(tuple(self.partitions_defs))
+
+    def get_partition_key(self, partition_key_by_dimension: Mapping[str, str]) -> MultiPartitionKey:
```

Comment:
I found this function a tiny bit confusing, because it led me to believe that there was some attribute of the `MultiPartitionsDefinition` that was required to build the `MultiPartitionKey`. If I understand correctly, the purpose of this function is just validation though? If that's the case, it might be preferable to change it into a function named something like `check_matches_partition_dimensions` that accepts a `MultiPartitionKey`? Or could consider leaving it out given that, in the only place this is used, it's us, not the user, who's supplying the arguments?

Reply:
That makes sense, I've removed this. If they wanted to, users can use the `MultiPartitionsKey` constructor instead to create a partition key